gobblin-runtime/src/main/java/org/apache/gobblin/runtime/messaging/DynamicWorkUnitConsumer.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.gobblin.runtime.messaging;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import org.apache.gobblin.runtime.messaging.data.DynamicWorkUnitMessage;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Receives {@link DynamicWorkUnitMessage} sent by {@link DynamicWorkUnitProducer}.
+ * The class is responsible for fetching the messages from {@link MessageBuffer}. All business logic
+ * is done in the {@link DynamicWorkUnitMessage.Handler}.<br><br>
+ *
+ * This consumer can be used to poll a message buffer (e.g. HDFS or Kafka) using
+ * {@link ScheduledExecutorService#scheduleAtFixedRate(Runnable, long, long, TimeUnit)} to call the
+ * {@link Runnable#run()} method periodically in a background thread <br><br>
+ *
+ * Each new {@link DynamicWorkUnitMessage} is passed to a {@link DynamicWorkUnitMessage.Handler}
+ * and will call {@link DynamicWorkUnitMessage.Handler#handle(DynamicWorkUnitMessage)}
+ */
+public class DynamicWorkUnitConsumer implements Runnable {
+  private static final Logger LOG = LoggerFactory.getLogger(DynamicWorkUnitConsumer.class);
+  protected MessageBuffer<DynamicWorkUnitMessage> messageBuffer;
+  protected List<DynamicWorkUnitMessage.Handler> messageHandlers;
+
+  public DynamicWorkUnitConsumer(
+      MessageBuffer<DynamicWorkUnitMessage> messageBuffer,
+      Collection<DynamicWorkUnitMessage.Handler> handlers) {
+    this.messageBuffer = messageBuffer;
+    this.messageHandlers = new ArrayList<DynamicWorkUnitMessage.Handler>(handlers);
+  }
+
+  /**
+   * Fetches all unread messages sent by {@link DynamicWorkUnitProducer} and
+   * calls {@link DynamicWorkUnitMessage.Handler#handle(DynamicWorkUnitMessage)} method for each handler added via
+   * {@link DynamicWorkUnitConsumer#DynamicWorkUnitConsumer(MessageBuffer, Collection)
+   */
+  public void run() {
```

Comment:
This run method will need to called in an infinite loop outside of the object. I'm thinking maybe using eventbus inside this consumer, and make the messageBuffer a runnable. Whenever the buffer receives a message, it post an event through the eventbus.
In this case, within the buffer, pull or push based consumption can be seamless for the consumer.